### PR TITLE
feat: Simple service templating

### LINKED ISSUE #48: Templating service files from ENV
### Intro
Okay, as a user i would like to be able to use variables inside service.toml files.
From my understanding there is no possibility to do that out of the box, i.e.
setup i have depends on various things, i.e. where the user has its own repo, what's user's username is, etc. etc.

### In the field:
imagine following service file:
```
....
command = "../../repos/service/target/debug/service"
user = "esavier"
....

```
i would love to be able to modify it to look like that:
```
....
command = "${WORKSPACE}/target/${TARGET}/service"
user = "${USER}"
....

```
of course this is the variable format used by bash, and it's only here to provide an example and a background. 

###  What i would like to achieve?
flexibility, right now i have to go through some steps that will ensure that path exists (like symlinking, bleeh)

### PR DESCRIPTION
**Pull request by esavier**
### Motivation and Context
Closes #48 

### Description
* Added simple preconfiguration stage when loading service files. Since service files are loaded into string, its easy to feed them to the templating engine.
* Added `Templar` as a templating engine 
* changed sample service file to use USER value (environment variable), was default. Does exactly the same what was done before, but shows potenatial of templating from ENV
* changed test to use user name instead defaulting to UID
* documentation related to change

### How Has This Been Tested?
* cargo test
* using templated service file (manual)

### Types of changes
- x New feature (non-breaking change which adds functionality)

### Checklist:
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

### CODE REVIEW

**Comment by FedericoPonzi** [comment]
Thanks a lot for this change, it looks very neat! 

Just to be extra sure, I would like to have an additional test for this. The test should:
* deserialize a template with a missing env variable, assert it fails.
* then you define the missing env variable, rerun the deserialization which should succeed this time. 

Should be straightforward, and after that I think we can merge this! :) 

Also, if you're up for it, I think it should be possible remove code duplication by calling `from_str` implementation from inside `from_file`.

**Comment by esavier** [comment]
i searched through the `templar` documentation, and it seems there is no straightforward way of determining if the template rendered correctly. Missing envs are just empty strings, since thats how libc interacts with its environment (linux way :D ). Technically this is ok, since no legacy code is broken, but:

I have the idea how we can build on this, but it will require configuration precheck phase i proposed in a different issue.
in short, a separate function that finds every env template and checks if its empty, then emits a warning.

For now i would like to propose: 

1. additional documentation with warning that if this feature is used, it may fail silently.
2. dumping configuration after rendering for user inspection.

Let me know what you think, meanwhile i will check your other comments :)

**Comment by FedericoPonzi** [comment]
> i searched through the templar documentation, and it seems there is no straightforward way of determining if the template rendered correctly.
In my mind, I was thinking about using [`std::env::set_var`](https://doc.rust-lang.org/std/env/fn.set_var.html) for a custom variable `ABC`, define a service using that variable, asserting that after applying templar, we have the expected resulting service. If this is not possible, then I'm ok with what you proposed! Many thanks!

**Comment by esavier** [comment]
Hey, sorry i am not able to make regular updates right now, but i want you to know i have a more elegant solution for this, without using the templating engine itself. I was searching for some way to respond with error if templating was performed with empty value, and i think i've got something that works quite nicely. I will update the pr soon :)
thank you for your patience

**Comment by FedericoPonzi** [comment]
No worries! Looking forward to the update, I'm very curious now!

**Comment by esavier** [comment]
Yay, okay its green :)
sorry it took so long, I tried few approaches, then vacation, broken pc and other stuff... but its finally there.
It should work out of the box, and there is proper error handling.

File: src/horust/formats/service.rs
Comment (FedericoPonzi) [inline src/horust/formats/service.rs:711]: I think this needs to be cleaned up

File: DOCUMENTATION.md
Comment (FedericoPonzi) [inline DOCUMENTATION.md:36]: It looks like the documentation needs to be updated accordingly

File: src/horust/formats/service.rs
Comment (FedericoPonzi) [inline src/horust/formats/service.rs:151]: This is not needed

File: example_services/templating.toml
Comment (FedericoPonzi) [inline example_services/templating.toml:5]: This also needs to be updated

**Review by FedericoPonzi — CHANGES_REQUESTED** [review]
I think it still needs a bit of cleanup! Anyway, I would like to understand what are the advantaged of using shellexpand vs templar?
* I find `"{{ env('USER') }}"` more self explanatory then `$USER`
* shellexpand is not currently being maintinaed: https://github.com/netvl/shellexpand
Thanks!

**Comment by esavier** [comment]
> I think it still needs a bit of cleanup! Anyway, I would like to understand what are the advantaged of using shellexpand vs templar?

Technically, `shellexpand` is not too outdated yet, and hopefully it will find a maintainer soon  (I believe the current maintainer's decision was very recent). It does not have any major dependencies, and that means it does not require much of maintaining, but sure, it may be a problem. Still it has 40k downloads monthly and from its class it's the best solution I found.
Anyway I can change the implementation to `envsubst` which is a little bit less popular, but still may work.

My reasoning was that with templar, environment templating was a side-effect, and not really a desired feature. Moreover, templar couldn't discern whenever environment existed or not, while shellexpand will fail, indicating configuration problem. Thanks to that I was able to do neat error-handling with a proper message, and that behavior is slightly less confusing in case of error.

**Comment by FedericoPonzi** [comment]
I was just questioning the decision of replacing templar to understand the reasoning, I don't currently find concerning shellexpand without maintainer (at this moment) so I'm ok with that. As soon as you update the docs, I think we can merge this! Thanks a lot for the effort!

**Comment by esavier** [comment]
Okay, sure, no problem, I just wanted to explain my decisions :)
Worst case scenario i was also ready to roll back the changes and work a bit more on templar.

I will try to fix commends today after work :+1:

File: example_services/templating.toml
Comment (esavier) [reply example_services/templating.toml]: Everything should be cleaned in the newest commit

File: DOCUMENTATION.md
Comment (FedericoPonzi) [inline DOCUMENTATION.md:18]: I guess this should be also updated.

Comment (FedericoPonzi) [reply DOCUMENTATION.md:18]: Not sure if it would be worth it out a link to the crate's doc, in order to allow users to understand which features are supported... WDYT?

Comment (esavier) [reply DOCUMENTATION.md]: Should be okay, 
I also forgot about one of the files, I noticed during manual testing.
Please resolve the comment if you are satisfied with the result :)

**Comment by FedericoPonzi** [comment]
I'm very happy about the final result! Thanks a ton for the contribution!

**Pull request merged**
